### ISSUE TITLE
Questions about shorthand font

### ISSUE THREAD

**Issue by flamenco108** [dscorbett/duployan-font#8]
Hi David,
I'm trying to contact somebody that could give me some advices of creating stenographic (shorthand) fonts. I have found you by your GitHub avatar and  it turned out that you work on Duployan font. As I see, you are experienced programmer (which I'm not) and skilled in the font-world (which I'm not). May I ask you several questions?

(Several months ago I downloaded this font and tried to use, but apparently it didn't work as I expexted - just now I read that perhaps it does not work in common text editor)

I work on stenographic font for Polish shorthand system. But from the beginning I have an idea that it must be usable just in eg. LibreOffice in a straight way - user types and sees the stenogramm.

In my first approach that I made ca. 10 years ago I used Fontforge to design shapes and glyphs and then I used the SIL technology to code all the dependencies between glyphs I could code. Which means - not so much. And also it was limited to text editors that could render the SIL (eg. Word didn't). Then I put this project away and about half a year ago I returned.

Years have passed and apparently the whole OpenType technology seems to be mature enough to use it instead of SIL. Eventually I can use SIL where OpenType would be insufficient.

Then the font might be consisted of:

- glyphs and shapes with incoming and outgoing anchors to interconnect;
- glyphs and shapes variants as above for combinations;
- multiple .fea files with lots of rules steering the process;
- SIL files to do the rest if need be.

So my questions are:

1. Do you think that this way is possible to achieve the usable result?
2. Which approach would be better: desing all glyphs and variants and then just use if->then - or - create variants by glueing together shapes on the fly using lookups and features?
3. I still look for approach to find a way to variant shorthand techniques (like write shorter or longer) - but to be used easily in text editor. My first idea is to create several font files like Normal, Italic, Bold - but of course they would be not italic, but shorter, not bold but even shorter... But I don't know if it might work. What is your advice?
4. I'm a beginner in Python but I see that it might be useful to automate work (I will have to get into it soon as there are no preview of anchors in Fontforge, so I will have to generate font and preview it in ready text file). I see that in Duployan font you use just Python. Do you think, should I rewrite all the project into Python that would be the keeper of all the process of font generation?

I hope you will find some time to help the amateur

If you prefer to answer via email, here is mine: flamenco108@stenografia.pl

Best regards

Krzysztof Smirnow (Stenografow)

**Comment by flamenco108** [comment]
Thanks for deepened answer.
1. Users are supposed just to type. Or Copy-Paste. And the result should be the proper shorthand writing.
2. My goal is three-fold:
2.1. Training of reading (in shorthand it's easier to write than to read);
2.2. Full and detailed documentation of the shorthand system (what could be better than the scripting/programming language?);
  2.2.1. Working on this still I find a lot of incoherrences that I have to solve, which is good;
2.3. In the future: training of some kind of expert system to read the shorthand and transcribe it to latin text on the fly.

I have limited understanding of the language of OpenType and especially of its limitations. I suppose it may affect my project somewhere in the future and I'm a bit afraid of that. I'm learning scripting to avoid possible need of manual rewriting large chunks of code.

*Ste***Mi** is similar to Gabelsberger or Stolze (or Sheithauer) shorthands - consonants are basically going down and vowels basically going up. But there are also a lot (hundreds in fact) of consonant groups in Polish, so I had to design a kind of spontanious way to create them, to avoid necessity to learn too many forms. But in the font they also must be somehow drawn. And here I want to use simpler parts of strokes. But I don't know if it would work. I want to code them in .fea file.

Example:

`k` is a vertical high stem, *(and alone it is just that)*

and `ka/ke/ko/ku` etc are `k-stem+round_ending+vowel`

but `kia/kie/kio/kiu` etc are `k-stem+sharp_ending+vowel`

Mentioned endings I can reuse also in `p,b,g,h,m,n` etc letters. But to have a good look of it I need also variants of them *(another for consonant+a/o, other for consonant+e etc)*.

Similar way I want to create consonant groups and so on.

If I add anchors to each simpler part, I hope to be able to create such connections by features and lookups. Am I right? (That was a question I wanted to ask ;-)) Otherwise I will have to create a lot of ligatures and just substitute by them - but when I will have to make a minor correction in some part of stroke, I will have to manually correct it in each ligature, which I want to avoid.

So now I have eg. 22 forms of `a` vowel (inclined on 4 angles, longer and shorter, and 4 types of anchor placing). Of course they are basically references to basic shapes. And rules in .fea file must manage it.

So I repeat my question: **can I count on proper work of anchors that would create the final letters?**

Example:

> `Szwecja` (Sweden) = `upper arch for bevels + capital letter marker + stem inclined 20 degrees to the left + right-handed loop inclined as above + semi-horizontal stem (e) + right-handed spiral (cj) + stem inclined 30 degrees to the right (basic a)`

or

> `Szwecja` = `Sz_w + semi-horizontal stem (e) + right-handed spiral (cj) + stem inclined 30 degrees to the right (basic a)`

Would the first example work?

What is the best practice in the world of font creators?

### About steno-generators


I learned the first address (clausthal) many years ago. He in fact inspired me to work on computer-generated shorthand. BTW I can give you another [address to learn of stenogenerators](https://www.jens-wawrczeck.de/stenogenerator/).

I've spent a lot of time discussing above topic with some Polish programmer. I'm not a programmer, so that's why I decided to go the way of fonts. And look where I am: scripts and programs all around. ;-)

### About using italic and bold

I just look for a way to keep these fonts in one family to be used everywhere similar way. But they don't differ by their thickness or slope, they differ by the shortness of writing, and how much sigilla (symbols for words, phrases or even sentences) are used.

**Comment by dscorbett** [comment]
Your idea of connecting simple glyphs with cursive anchors should work.

I can’t say which approach is better in your “Szwecja” example. They are probably both fine. OpenType shorthand fonts are not common so there isn’t really any best practice. The only problem I can think of is that the more glyphs there are, the more opportunities there are for rounding errors in glyph positioning, which can make the pixels around the connection point appear slightly darker because the overlap isn’t perfect. That is a minor point in favor of your second “Szwecja” implementation. You’ll have to weigh the pros and cons based on your own experience, but if you aren’t sure which to pick, it probably doesn’t matter much.

You may have to test various approaches but the hard part, I think, will be systematically analyzing the shorthand system and how all its letters join with each other. It sounds like you’re doing that anyway. Once that is done, it should be straightforward to translate those rules into OpenType feature rule syntax. For example:

```afdko
feature rclt {
    sub [k g] a' by round.ka a;
    sub [k g] e' by round.ke e;
    sub [k g] o' by round.ko o;
    sub [k g] u' by round.ko u;
} rclt;

feature rclt {
    sub [k g] i' a' by a;
    sub [k g] i' e' by e;
    sub [k g] i' o' by io;
    sub [k g] i' u' by iu;
} rclt;
```

That uses glyphs like `round.ka` for the small rounded connection between `k` and `a`. Multiple such glyphs are needed because different consonant–vowel pairs form different angles. This would let you use the same `a` glyph after any consonant, hard or soft, only changing the rounded glyph. This is not the only way to do it and might not even be the best way.

Before you get too far, though, I recommend making a simple test font that only supports a few syllables. See if it works in Word (or whatever editors you plan to support). In particular, Word has bad OpenType support; it would not surprise me if it doesn’t support left-to-right cursive attachment for the Latin script (which I assume is your proposed implementation). It might work, but if it doesn’t, it is better to know as soon as possible.

**Comment by flamenco108** [comment]
You gave me a lot of clues to rethink, thanks very much. It is sad that I cannot continue with the main works of the project, but apparently I have to first  rework and reorder all my environment: 

1. I created test-file as you advised.
2. But to test, I need already scripts that generate wanted font, include fea-files and so on - which I planned to work on much later.

But from this test file I already learned that it is not so bad with overlapping connections of glyphs. 
(BTW. I've read about such a problem in memoirs of one Polish programmer that worked for Sun in '80-s on the library which perhaps later became the OpenGL. He had a problem with black dot in the centre of the red circle. As there were no elegant solution, he had to work around it.)

### Simple shapes connected at runtime with cursive rules

In fact it was the reason why I decided to recreate the font from the very beginning and not just reuse my old sfd files. I had an idea that even very cursive writing could be embraced into smaller number of simpler geometrical representants. Having such a resource I could create another font for SteMi just slightly modyfying these basic shapes and after that doing minor adjustments.

From this comes my next question: how important is to not to have overlaps and crossing splines in the glyph (in fact I also don't understand the importance of extremes - there are so many tools that I don't even have a clue, what they mean...)? Fontforge cries a lot about it when I generate font from GUI. But the font appears and works. So, what is all about?

**Comment by flamenco108** [comment]
I have quick question about cursive anchors:
What's your advice:
I want to keep cursive anchors in separate FEA file to easily manage them (there will be several hundreds of them). In the generation/testing process I want to put them into the SFD fontforge file just before font generation. But I need them in the SFD (same as I need this file before it's filled with lookups and features, just shapes), as perhaps there will be some issues and then I need to make some corrections before font generation. And fontforge gui is not very friendly for me as far as anchors are concerned. I mean,  it's very friendly to establish these anchors, but not so to manage them.

So, following your advice, I have written/found in net several scripts:

- export anchors (and other features) to .FEA file - to get basic schema of the OpenType functional notation;
- import FEA file to SFD file;
- etc. off this topic.

The script importing features does it well. But not with the part with cursive anchors. Let me show 
This is the importing function:
```
def add_fea_to_sfd(input_sfd, input_fea, output_sfd):
    try:
        font = fontforge.open(input_sfd)
        font.mergeFeature(input_fea)
        font.save(output_sfd)
        font.close()
        print(f"Plik .fea został pomyślnie dodany do pliku {input_sfd} i zapisany jako {output_sfd}.")
    except Exception as e:
        print(f"Wystąpił błąd podczas dodawania pliku .fea do pliku .sfd: {e}")
```


The snippet of the main file and the FEA file with anchors:

```
languagesystem DFLT dflt;


# GPOS 


#Anchors anchory kotwice
include(tst01-anchors.fea);

# Mark attachment classes (defined in GDEF, used in lookupflags)
@GDEF_Simple = [\.null \uni000A \space (...) \gorOMluk \dolMOluk ];

table GDEF {
  GlyphClassDef @GDEF_Simple, , , ;
} GDEF;

# GSUB

lookup litery {
  #sub k' by stemK dolTW;  # Correct: Single quotes around glyph names
  sub g' s' by stemG dolTW koloS;
  #sub k' e' by stemK dolTW eprost0;
  #sub g' z' by stemG dolMK30 koloZ;
} litery;

feature rclt {
    lookup litery;
    #sub k'  by stemK dolTW;
    #sub g' s' by stemG dolTW koloS;
    #sub k' e' by stemK dolTW eprost0;
    #sub g' z' by stemG dolMK30 koloZ;
} rclt;
```
and just a fragment of tst01-anchors.fea:
```
lookup curs-anchory {
  lookupflag 0;
    pos cursive \stemK <anchor 60 1649> <anchor 60 60>;
    pos cursive \gorP.brzh <anchor 60 919> <anchor 60 839>;
    pos cursive \stemI.krt <anchor 60 282> <anchor 60 60>;
    pos cursive \stemKRT <anchor 60 180> <anchor 60 60>;
(...)
} curs-anchory;

feature curs {
 script DFLT;
     language dflt ;
     lookup curs-anchory;
} curs;
```
And on the import I get strange error:
```
Expected '{' in feature definition on line 1 of tst01-anchors.fea
Expected '{' in feature definition on line 78 of tst01-anchors.fea
There were errors when parsing the feature file and the features have not been applied
```
It looks like bug, but I'm not programmer and I can't be sure.

But anyway: Maybe you have an idea, what's wrong with it and if nothing, how experienced font creators keep their cursive anchors to easily manage them, if there is so big mess with importing them?

Thanks for your help I hope I'm not too annoying.

**Comment by dscorbett** [comment]
A good library for dealing with feature files is fontTools. It can give you a second opinion if FontForge reports an error in a feature file.
```python
from fontTools.feaLib.parser import Parser
Parser('tst01-anchors.fea').parse()
```
In this case, my guess is FontForge doesn’t accept a hyphen in a lookup name like `curs-anchory`.

Everything in my Duployan font is generated in Python, so I only have one choice for where to keep cursive anchors, so I can’t really say what an experienced font creator would do in your situation. I expect it would probably be easier to keep them in the SFD. Otherwise, every time you modify a glyph’s contours, you have to modify its anchors in a separate text file without any realtime visual feedback.

**Comment by flamenco108** [comment]
Here was a long epistle, but finally I decided to struggle with all these uncertainties by myself - at least until I will hit the wall. 

(I was able to start only `liga` feature, the rest don't work and I don't know why yet. But I understand that basic technique for me will be the *chaining contextual substitutions*, so there's long way to go before me.) 

So let me just present below a short graphic presentation, how I have planned to deal with shorthand letters in my case (long time ago I became attached to the idea of a geometric approach to glyph building) - by the example of letter `n` combined with `o` and softened/voiced by `i` (glyph names are contractual):

`n` is combined of 2 small arcs (`arc.begin`+`arc.end`), but softened second arc changes to `soft.end` and i-voiced to `i.45`:

![nnini](https://github.com/dscorbett/duployan-font/assets/5072952/63cc6dd1-0203-4070-9238-906034c5d16e)

`n   ń   ni`

Combined with `o` first `arc.begin` (*on*) changes into `on.begin`; when it's medial form (when therefore attaches to eg. `k`); `arc.end` disappeares in final form; the `arc.end` changes into `no.end` when it's *no*; and into `hard.end` when it's *non*:

![ononnoononononon](https://github.com/dscorbett/duployan-font/assets/5072952/2eb8b89c-4c5d-43bf-aca9-40f6db77ec1e)

`on(medial)  on(final)  no   ono   non   onon`

So, when `n`+`o` combinations are softened/voiced by `i`, when *on* becomes *oń*, the `arc.eng` changes into `soft.end`, when *oni* into `i.45`... And then, when *nio*, `arc.end`->`soft.end`+`o.45.alt` (in contrast to *no* as above), and so on.


![onioninionionniononiononion](https://github.com/dscorbett/duployan-font/assets/5072952/b4b38610-7369-403d-81dd-2d47cb986d8a)

`oń  oni  nio  nion  nion(final) onion(medial) onion(final`

And this is just one, simple example. Everywhere there are also modifications (like different angles for eg.) depending on the next letter attached (filius ante patrem) and sometimes previous (pater ante filium).

![shapes](https://github.com/dscorbett/duployan-font/assets/5072952/ec4fbbce-4b27-487c-a142-fdbc08fc33aa)

I just ask myself (as I have shown you just the simplest shapes-glyph parts), what  ligatures MUST be created and what could be generated by .fea code?

Well, today nothing works, so it's still open question.

**Comment by Zireael07** [comment]
@flamenco108 Hi, I love your stenography/stenotype site.

As you said you are not too good with font files and formats, another approach I've thought of (person working on a custom system too) would be to output plain SVG.

https://github.com/inconvergent/gridfont or Hershey fonts are two ways of describing your basic shapes that can be converted to SVG

**Comment by flamenco108** [comment]
Thank you for kind words. I haven't answered for several days because I didn't know if it's OK to talk on someone's github-channel off-topic. 

So I want to encourage you to write to me on email flamenco108@stenografia.pl, as I don't know if David would like to have such conversation here...

Anyway I checked this link and I find it really interesting. Though I suppose it is not enough complicated to make possible to draw shorthand. My intention is to create a font - that anybody, not just people more familiar with IT, can try shorthand. And the same way it would be the perfect documentary of the system. And I can learn something new.

Since then I have moved forward with my knowledge of the *fea* language and of course, everything must be rebuilt from the very beginning. Everything is new for me here...

Thenk you ;-)
